src/runtime_src/core/tools/xbflash2/OO_Program_Spi.cpp

```diff
@@ -0,0 +1,220 @@
+        if (pri.fail())
+            return -EINVAL;
+        ret = xspi.xclUpgradeFirmware1(pri);
+    }
+    else {
+        firmwareImage pri(primary_file[0].c_str());
+        firmwareImage sec(primary_file[1].c_str());
+        if (pri.fail() || sec.fail())
+            return -EINVAL;
+        ret = xspi.xclUpgradeFirmware2(pri, sec);
+    }
+    return ret;
+}
+
+int
+spiCommand(po::variables_map vm) {
+    std::string sBar, sBarOffset;
+    size_t baroff = INVALID_OFFSET;
+    int bar = 0;
+    int err = -EINVAL;
+
+    //optional command line args
+    try {
+        sBar = vm["bar"].as<std::string>();
+        if (!sBar.empty())
+            bar = std::stoi(sBar);
+        sBarOffset = vm["bar-offset"].as<std::string>();
+        if (!sBarOffset.empty()) {
+            std::stringstream sstream(sBarOffset);
+            sstream >> baroff;
+        }
+    }
+    catch (...) {
+    }
+
+    if (vm.count("revert-to-golden")) {
+        err = reset(vm, bar, baroff);
+    }
+    else if (vm.count("image")) {
+        err = flash(vm, bar, baroff);
+    }
+
+    return err;
+}
+} //end namespace 
+
+// ----- C L A S S   M E T H O D S -------------------------------------------
+
+OO_Program_Spi::OO_Program_Spi( const std::string &_longName, bool _isHidden )
+    : OptionOptions(_longName, _isHidden, "Flash type - spi")
+    , m_help(false)
+{
+
+  m_optionsDescription.add_options()
+      ("device,d", po::value<std::string>(), "The \"Bus:Device.Function\" (e.g., 0000:d8:00.0) device of interest.\n")
+      ("dual-flash,u", "Specifies if the card is dual flash supported.\n")
+      ("bar,b", po::value<std::string>(), "BAR-index, default is 0.\n")
+      ("bar-offset,s", po::value<std::string>(), "BAR-offset-for-QSPIPS, default is 0x40000.\n")
+      ("image,i", po::value<std::vector<std::string>>(), "Specifies MCS or BOOT.BIN image path to update the persistent device.\n")
+      ("revert-to-golden,r", "Resets the FPGA PROM back to the factory image.\n")
+      ("help", boost::program_options::bool_switch(&m_help), "Help to use this sub-command")
+  ;
+}
+
+void
+OO_Program_Spi::execute(const SubCmdOptions& _options) const
+{
+  XBU::verbose("SubCommand option: Flash type - spi");
+  XBU::verbose("Option(s):");
+  for (auto & aString : _options)
+    XBU::verbose(std::string(" ") + aString);
+
+  // Honor help option first
+  if (std::find(_options.begin(), _options.end(), "--help") != _options.end()) {
+    printHelp();
+    return;
+  }
+
+  // Parse sub-command ...
+  po::variables_map vm;
+
+  try {
+    po::store(po::command_line_parser(_options).options(m_optionsDescription).run(), vm);
+    po::notify(vm); // Can throw
+  }
+  catch (po::error& e) {
+    std::cerr << "ERROR: " << e.what() << "\n\n";
+    printHelp();
+    throw std::errc::operation_canceled;
+  }
+
+  if(m_help) {
+    printHelp();
+    throw std::errc::operation_canceled;
+  }
+
+  XBFU::sudo_or_throw();
+ 
+  if (spiCommand(vm) == -EINVAL)
```

Comment:
instead of C style error, why not use std::error

Reply:
comments addressed.